certora/specs/Gates.spec

```
@@ -0,0 +1,134 @@
+}{
+    require(currentContract.sharesGate != 0, "require gating to be enabled for shares");
+
+    // Trick to require that all the following addresses are different.
+    require(MorphoMarketV1Adapter == 0x10, "ack");
+    require(MorphoVaultV1Adapter == 0x11, "ack");
+    require(currentContract == 0x12, "ack");
+
+    require (!canReceiveShares(user), "require that the user under scrutiny can't receive shares");
+
+    uint256 sharesBefore = balanceOf(user);
+
+    f(e, args);
+
+    assert balanceOf(user) <= sharesBefore;
+}
+
+rule cantSendShares(env e, method f, calldataarg args, address user, uint256 shares)  filtered {
+    f -> f.selector != sig:MorphoMarketV1Adapter.skim(address).selector &&
+         f.selector != sig:MorphoVaultV1Adapter.skim(address).selector
+}{
+    require(currentContract.sharesGate != 0, "require gating to be enabled for shares");
+
+    // Trick to require that all the following addresses are different.
+    require(MorphoMarketV1Adapter == 0x10, "ack");
+    require(MorphoVaultV1Adapter == 0x11, "ack");
+    require(currentContract == 0x12, "ack");
+
+    require (!canSendShares(user), "require that the user under scrutiny can't send shares");
+
+    uint256 sharesBefore = balanceOf(user);
+
+    f(e, args);
+
+    assert balanceOf(user) >= sharesBefore;
+}
+
+rule cantReceiveAssets(env e, method f, calldataarg args, address user)  filtered {
+    f -> f.selector != sig:MorphoMarketV1Adapter.skim(address).selector &&
+         f.selector != sig:MorphoVaultV1Adapter.skim(address).selector
+}{
+    require(currentContract.receiveAssetsGate != 0, "require gating to be enabled for receiving assets");
+
+    require(asset() == ERC20, "ack");
+
+    // Trick to require that all the following addresses are different.
+    require(MorphoMarketV1Adapter == 0x10, "ack");
+    require(MorphoVaultV1Adapter == 0x11, "ack");
+    require(currentContract == 0x12, "ack");
+    require(asset() == 0x13, "ack");
+
+    require (user != MorphoMarketV1Adapter && user != MorphoVaultV1Adapter && user != currentContract, "require that the vault and the adapters are allowed to receive assets");
+    require (currentContract.liquidityAdapter == 0x0 || currentContract.liquidityAdapter == MorphoMarketV1Adapter || currentContract.liquidityAdapter == MorphoVaultV1Adapter, "require that the liquidity adapter is unset or a known implementation");
+
+    require (!canReceiveAssets(user), "assume that the user under scrutiny can't receive assets");
+    require (ghostBalanceChangeAllowed[user] == true, "setup the ghost state");
+
+    f(e, args);
+
+    assert ghostBalanceChangeAllowed[user];
+}
+
+rule cantSendAssets(env e, method f, calldataarg args, address user)  filtered {
+    f -> f.selector != sig:MorphoMarketV1Adapter.skim(address).selector &&
+         f.selector != sig:MorphoVaultV1Adapter.skim(address).selector
+}{
+    require(currentContract.sendAssetsGate != 0, "require gating to be enabled for sending assets");
```

```suggestion
    require(currentContract.sendAssetsGate != 0, "setup gating");
    require (!canSendAssets(user), "setup gating");
```
and remove the corresponding line below. This explains why we do those requires: we place ourselves in a scenario where a user is gated, and check that the corresponding change cannot be made